### ISSUE TITLE
CanDeactivate not triggered when navigating from lazy loaded module component to root module

### ISSUE THREAD

**Issue by HighSoftWare96** [angular/angular#28730]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core, @angular/router.

### Is this a regression?

 I do not know precisely. There was maybe a similar issue described here: #16868

### Description

 
When I put a *CanDeactivate guard* for the route of a *component* that is included inside a *lazy loaded module* and I try to navigate (for example with the back button) from this route *to a route that is included in the root module* (app.module.ts, for example the homepage) the **CanDeactivate guard it is not called**. This behaviour does not appear when navigating between two components (in which at least one is protected by the CanDeactivate) that are included in the same lazy-loaded module or are included in two different lazy-loaded modules. 

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-issue-deactivate-guard-lazyload-module

## 🔥 Exception or Error
There are no exceptions or errors sadly...

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 7.2.2
Node: 10.15.0
OS: darwin x64
Angular: 7.2.1
... animations, cdk, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.12.2
@angular-devkit/build-angular     0.12.2
@angular-devkit/build-optimizer   0.12.2
@angular-devkit/build-webpack     0.12.2
@angular-devkit/core              7.2.2
@angular-devkit/schematics        7.2.2
@angular/cli                      7.2.2
@angular/flex-layout              7.0.0-beta.23
@angular/http                     7.2.2
@ngtools/webpack                  7.2.2
@schematics/angular               7.2.2
@schematics/update                0.12.2
rxjs                              6.3.3
typescript                        3.2.2
webpack                           4.28.4
</code></pre>

**Comment by lxieyang** [comment]
Any workaround/news on this issue?

**Comment by javaHelper** [comment]
How can we solve this issue ? Any pointers please ?

**Comment by javaHelper** [comment]
https://stackoverflow.com/questions/58240342/angular-candeactivate-not-working-and-value-showing-null - Please guide here

**Comment by csutorasr** [comment]
> https://stackoverflow.com/questions/58240342/angular-candeactivate-not-working-and-value-showing-null - Please guide here

This is not a solution just a workaround.

**Comment by thakurarun** [comment]
I'm facing similar issue. Any fix or workaround?

**Comment by nhantt10** [comment]
Try implement window unload event handler. I think the guard is activated only when a navigation happens (link click, `router.navigate()`). For the browser-based navigation (back button, refresh button, home button), the guard won't be triggered. Instead, handle the window unload event will do the trick.

I tried these code in your stackblitz and it worked:
```typescript
export class TestComponent {
  // your code goes here
  @HostListener("window:beforeunload", ["event"])
  unloadHandler(event: WindowEventHandlers): boolean {
    alert("unloading");
    return false;
  }
}
```

**Comment by ThanhNhan776** [comment]
Try implement window unload event handler. I think the guard is activated only when a navigation happens (link click, `router.navigate()`). For the browser-based navigation (back button, refresh button, home button), the guard won't be triggered. Instead, handle the window unload event will do the trick.

I tried these code in your stackblitz and it worked:
```typescript
export class TestComponent {
  // your code goes here
  @HostListener("window:beforeunload", ["event"])
  unloadHandler(event: WindowEventHandlers): boolean {
    alert("unloading");
    return false;
  }
}
```

**Comment by narendrasinghrathore** [comment]
````
ERROR Error: Uncaught (in promise): TypeError: component.confirmAndNavigate is not a function
````
**When returning from lazy loaded module to root module.** 

Root Module


Lazy Loaded Feature Module 
  - Child component having unsaved changes, trigger canDeacticvate but provider don't have component reference

 - Child lazy loaded  modules having single components with canDeactivate guard working fine in between component navigation

**Working ( with canDeactivate guard )**
 - **/root-route/lazy-loaded-parent/lazy-loaded-child1** to **/root-route/lazy-loaded-parent/lazy-loaded-child2**

**Not working ( with canDeactivate guard )**
- From **/root-route/lazy-oaded-parent/lazy-loaded-child-xx** to **/root-route**

**Comment by flensrocker** [comment]
I got hit by this today, but not with the back-button of the browser but with a routerLink in a sidenav.

If I start the reproduction sample above, a reload of the app ist triggered when navigating back. If I open the preview in a new window the deactivate-guard is called on browser's back button.

Maybe I have another problem - but it seems so close: I have a deactivate-guard in an external package, import that module in the app.module and lazy-module, decorated the component and route in the lazy module and it won't get called, if I navigate back to the home component...

I have to investigate this further...

**Comment by ankushadd123** [comment]
I got stuck in the same situation today.

I have a nav-bar in a component. I have implemented separate child route/component for each tab of my nav-bar. I want to implement canDeactivate on one of my child component/ nav-bar tab.

Between tabs of nav-bar, it works fine but when I try to go to home from my child tab(implementing canDeactivate guard), it causes error. The value of child component is null inside canDeactivate guard.

Anyone knows how to fix this?
